src/consensus/aft/raft.h

```diff
@@ -2104,31 +2115,55 @@ namespace aft
         sort(match.begin(), match.end());
         auto confirmed = match.at((match.size() - 1) / 2);
 
-        if (confirmed < new_commit_idx)
+        if (
+          !new_agreement_index.has_value() ||
+          confirmed < new_agreement_index.value())
         {
-          new_commit_idx = confirmed;
+          new_agreement_index = confirmed;
         }
       }
-      RAFT_DEBUG_FMT(
-        "In update_commit, new_commit_idx: {}, "
-        "last_idx: {}",
-        new_commit_idx,
-        state->last_idx);
 
-      if (new_commit_idx != std::numeric_limits<Index>::max())
+      if (new_agreement_index.has_value())
       {
-        state->watermark_idx = new_commit_idx;
-      }
+        const auto new_commit_idx =
+          find_best_committable_index(new_agreement_index.value());
 
-      if (get_commit_watermark_idx() > state->last_idx)
-      {
-        throw std::logic_error(
-          "Followers appear to have later match indices than leader");
-      }
+        if (new_commit_idx.has_value())
+        {
+          if (new_commit_idx.value() > state->last_idx)
+          {
+            throw std::logic_error(
+              "Followers appear to have later match indices than leader");
+          }
+
+          RAFT_DEBUG_FMT(
+            "In update_commit, new_commit_idx: {}, "
+            "last_idx: {}",
+            new_commit_idx.value(),
+            state->last_idx);
 
-      commit_if_possible(get_commit_watermark_idx());
+          const auto term_of_new = get_term_internal(new_commit_idx.value());
+          if (term_of_new == state->current_view)
+          {
+            commit(new_commit_idx.value());
+          }
+          else
+          {
+            RAFT_DEBUG_FMT(
+              "Agreed index ({}), resulted in proposed commit index {}, which "
+              "is in term {}. Waiting for agreement on committable entry in "
```

Comment:
Perhaps "ack quorum" is clearer in the log, without a prior definition of agreement? idk.